File: patterns/_tests/BehavioralPatterns/Interpreter/java/integration.test.java
// ============================================================
//  Integration Tests — Interpreter: Arithmetic Expression Evaluator (Java)
//
//  Run: javac java.java integration.test.java && java org.junit.runner.JUnitCore InterpreterIntegrationTests
// ============================================================

import org.junit.Test;
import static org.junit.Assert.*;
import java.util.NoSuchElementException;

public class InterpreterIntegrationTests {

    // ── Full expression evaluation end-to-end ─────────────────
    @Test
    public void threeLevelNestedTreeEvaluatesCorrectly() {
        // ((a + b) * (a - b)) / 5  with a=10, b=5  →  75/5 = 15
        Expression a   = new VariableExpression("a");
        Expression b   = new VariableExpression("b");
        Context ctx    = new Context().set("a", 10).set("b", 5);

        Expression expr = new DivideExpression(
            new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b)),
            new NumberExpression(5));

        assertEquals(15.0, expr.interpret(ctx), 0.001);
    }

    @Test
    public void treeReuseThreeContexts() {
        Expression a    = new VariableExpression("a");
        Expression b    = new VariableExpression("b");
        Expression expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));

        assertEquals(75.0, expr.interpret(new Context().set("a", 10).set("b", 5)), 0.001);
        assertEquals(32.0, expr.interpret(new Context().set("a",  6).set("b", 2)), 0.001);
        assertEquals( 8.0, expr.interpret(new Context().set("a",  3).set("b", 1)), 0.001);
    }

    @Test
    public void displayProducesFullyParenthesisedString() {
        Expression expr = new MultiplyExpression(
            new AddExpression(new VariableExpression("a"), new VariableExpression("b")),
            new SubtractExpression(new VariableExpression("a"), new VariableExpression("b")));

        String s = expr.display();
        assertTrue(s.contains("(a + b)"));
        assertTrue(s.contains("(a - b)"));
        assertTrue(s.startsWith("("));
    }


    // ── Error propagation ─────────────────────────────────────
    @Test(expected = NoSuchElementException.class)
    public void missingVariableInDeepSubtreeSurfaces() {
        Expression expr = new AddExpression(
            new NumberExpression(1),
            new MultiplyExpression(new VariableExpression("missing"), new NumberExpression(2)));
        expr.interpret(new Context());
    }

    @Test(expected = ArithmeticException.class)
    public void divisionByZeroInSubtreeSurfaces() {
        Expression expr = new AddExpression(
            new NumberExpression(1),
            new DivideExpression(new NumberExpression(4), new NumberExpression(0)));
        expr.interpret(new Context());
    }


    // ── Context independence ───────────────────────────────────
    @Test
    public void parallelEvaluationsDoNotInterfere() {
        Expression a    = new VariableExpression("a");
        Expression b    = new VariableExpression("b");
        Expression expr = new AddExpression(a, b);

        Context ctx1 = new Context().set("a", 1).set("b", 2);
        Context ctx2 = new Context().set("a", 100).set("b", 200);

        assertEquals(  3.0, expr.interpret(ctx1), 0.001);
        assertEquals(300.0, expr.interpret(ctx2), 0.001);
    }


    // ── Arithmetic correctness ─────────────────────────────────
    @Test
    public void additionIsCommutative() {
        Context ctx = new Context().set("a", 3).set("b", 7);
        Expression ab = new AddExpression(new VariableExpression("a"), new VariableExpression("b"));
        Expression ba = new AddExpression(new VariableExpression("b"), new VariableExpression("a"));
        assertEquals(ab.interpret(ctx), ba.interpret(ctx), 0.001);
    }

    @Test
    public void multiplyByZeroGivesZero() {
        Context ctx  = new Context().set("a", 999);
        Expression expr = new MultiplyExpression(new VariableExpression("a"), new NumberExpression(0));
        assertEquals(0.0, expr.interpret(ctx), 0.001);
    }
}

File: patterns/_tests/BehavioralPatterns/Interpreter/csharp/integration.test.cs
// ============================================================
//  Integration Tests — Interpreter: Arithmetic Expression Evaluator (C#)
//
//  Run: dotnet test
// ============================================================

using Xunit;
using System;
using System.Collections.Generic;

public class IntegrationTests {

    // ── Full expression evaluation end-to-end ──────────────────
    [Fact]
    public void ThreeLevelNestedTreeEvaluatesCorrectly() {
        // ((a + b) * (a - b)) / 5  with a=10, b=5  →  75/5 = 15
        IExpression a   = new VariableExpression("a");
        IExpression b   = new VariableExpression("b");
        var ctx = new Context().Set("a", 10).Set("b", 5);

        var expr = new DivideExpression(
            new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b)),
            new NumberExpression(5));

        Assert.Equal(15, expr.Interpret(ctx));
    }

    [Fact]
    public void TreeReuseThreeContexts() {
        IExpression a    = new VariableExpression("a");
        IExpression b    = new VariableExpression("b");
        var expr = new MultiplyExpression(new AddExpression(a, b), new SubtractExpression(a, b));

        var cases = new[] {
            (a: 10.0, b: 5.0, expected: 75.0),
            (a:  6.0, b: 2.0, expected: 32.0),
            (a:  3.0, b: 1.0, expected:  8.0),
        };

        foreach (var (av, bv, expected) in cases) {
            var ctx = new Context().Set("a", av).Set("b", bv);
            Assert.Equal(expected, expr.Interpret(ctx));
        }
    }

    [Fact]
    public void DisplayProducesFullyParenthesisedString() {
        var expr = new MultiplyExpression(
            new AddExpression(new VariableExpression("a"), new VariableExpression("b")),
            new SubtractExpression(new VariableExpression("a"), new VariableExpression("b")));

        var s = expr.Display();
        Assert.Contains("(a + b)", s);
        Assert.Contains("(a - b)", s);
        Assert.StartsWith("(", s);
    }

    // ── Error propagation ──────────────────────────────────────
    [Fact]
    public void MissingVariableInDeepSubtreeSurfaces() {
        var expr = new AddExpression(
            new NumberExpression(1),
            new MultiplyExpression(new VariableExpression("missing"), new NumberExpression(2)));
        Assert.Throws<KeyNotFoundException>(() => expr.Interpret(new Context()));
    }

    [Fact]
    public void DivisionByZeroInSubtreeSurfaces() {
        var expr = new AddExpression(
            new NumberExpression(1),
            new DivideExpression(new NumberExpression(4), new NumberExpression(0)));
        Assert.Throws<DivideByZeroException>(() => expr.Interpret(new Context()));
    }

    // ── Context independence ───────────────────────────────────
    [Fact]
    public void ParallelEvaluationsDoNotInterfere() {
        IExpression a    = new VariableExpression("a");
        IExpression b    = new VariableExpression("b");
        var expr = new AddExpression(a, b);

        var ctx1 = new Context().Set("a", 1).Set("b", 2);
        var ctx2 = new Context().Set("a", 100).Set("b", 200);

        Assert.Equal(3,   expr.Interpret(ctx1));
        Assert.Equal(300, expr.Interpret(ctx2));
    }

    // ── Arithmetic correctness ─────────────────────────────────
    [Fact]
    public void AdditionIsCommutative() {
        var ctx = new Context().Set("a", 3).Set("b", 7);
        var ab  = new AddExpression(new VariableExpression("a"), new VariableExpression("b"));
        var ba  = new AddExpression(new VariableExpression("b"), new VariableExpression("a"));
        Assert.Equal(ab.Interpret(ctx), ba.Interpret(ctx));
    }

    [Fact]
    public void MultiplyByZeroGivesZero() {
        var ctx  = new Context().Set("a", 999);
        var expr = new MultiplyExpression(new VariableExpression("a"), new NumberExpression(0));
        Assert.Equal(0, expr.Interpret(ctx));
    }
}
